Running bestPracticesBaseline twice fails with "Required value was null."

It seems the serialization is not reversible?
```
Caused by: java.lang.IllegalStateException: Required value was null.
        at dev.zacsweers.moshix.sealed.reflect.MetadataMoshiSealedJsonAdapterFactory.create(MetadataMoshiSealedJsonAdapterFactory.kt:57)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:74)
        at com.squareup.moshi.CollectionJsonAdapter.newArrayListAdapter(CollectionJsonAdapter.java:54)
        at com.squareup.moshi.CollectionJsonAdapter$1.create(CollectionJsonAdapter.java:38)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:74)
        at com.autonomousapps.task.CheckBestPracticesTask$Action.execute(CheckBestPracticesTask.kt:228)
```

I'm running `gradlew :plugins:bestPracticesBaseline` or `gradlew :plugins:checkBestPractices`.

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -2,16 +2,11 @@
 java = "11"
 junit = "5.8.2"
 kotlin = "1.6.10"
-moshi = "1.12.0"
-moshix = "0.14.1"
+moshi = "1.14.0"
+moshix = "0.23.0"
 retrofit = "2.9.0"
 
 [libraries]
-agp = "com.android.tools.build:gradle:4.2.2"
-
-caffeine = "com.github.ben-manes.caffeine:caffeine:3.1.0"
-commons-io = "commons-io:commons-io:2.11.0"
-gradle-publish-plugin = "com.gradle.plugin-publish:com.gradle.plugin-publish.gradle.plugin:0.11.0"
 guava = "com.google.guava:guava:31.1-jre"
 
 javax-inject = "javax.inject:javax.inject:1"

--- a/plugin-best-practices-plugin/src/functionalTest/groovy/com/autonomousapps/FunctionalSpec.groovy
+++ b/plugin-best-practices-plugin/src/functionalTest/groovy/com/autonomousapps/FunctionalSpec.groovy
@@ -19,12 +19,19 @@ final class FunctionalSpec extends Specification {
     def project = new SimplePluginProject(tempDir, 'reporting')
 
     when:
-    buildAndFail(project.root, 'checkBestPractices')
+    def result = buildAndFail(project.root, 'checkBestPractices')
 
-    then:
-    project.report.text.trim() == project.expectedReport.trim()
+    then: 'Console output matches expected value'
+    result.output.contains project.expectedConsoleOutput
+
+    and: 'File version of console output matches expected value'
+    project.consoleReport.text.trim() == project.expectedConsoleReport.trim()
+
+    and: 'Json report matches expected value'
+    project.jsonReport.text.trim() == project.expectedJsonReport.trim()
   }
 
+  // Same as above, but using `check` lifecycle task instead.
   def "can check best practices with 'check' task"() {
     given:
     def project = new SimplePluginProject(tempDir)
@@ -33,7 +40,7 @@ final class FunctionalSpec extends Specification {
     buildAndFail(project.root, 'check')
 
     then:
-    project.report.text.trim() == project.expectedReport.trim()
+    project.consoleReport.text.trim() == project.expectedConsoleReport.trim()
   }
 
   def "can create best practices baseline"() {
@@ -44,6 +51,19 @@ final class FunctionalSpec extends Specification {
     build(project.root, 'bestPracticesBaseline')
 
     then:
-    project.bestPractices.text.trim() == project.expectedBaseline.trim()
+    project.baselineReport.text.trim() == project.expectedBaseline.trim()
+  }
+
+
+  def "respects baseline"() {
+    given:
+    def project = new SimplePluginProject(tempDir)
+
+    when: 'Generate the baseline and then check best practices'
+    build(project.root, 'bestPracticesBaseline')
+    build(project.root, 'checkBestPractices')
+
+    then: "Build doesn't fail"
+    project.baselineReport.text.trim() == project.expectedBaseline.trim()
   }
 }

--- a/plugin-best-practices-plugin/src/functionalTest/groovy/com/autonomousapps/fixture/SimplePluginProject.groovy
+++ b/plugin-best-practices-plugin/src/functionalTest/groovy/com/autonomousapps/fixture/SimplePluginProject.groovy
@@ -15,8 +15,9 @@ final class SimplePluginProject {
   }
 
   Path root = tempDir
-  Path report = root.resolve('build/reports/best-practices/report.txt')
-  Path bestPractices = root.resolve('best-practices-baseline.json')
+  Path consoleReport = root.resolve('build/reports/best-practices/report.txt')
+  Path jsonReport = root.resolve('build/reports/best-practices/report.json')
+  Path baselineReport = root.resolve('best-practices-baseline.json')
 
   private void build() {
     newFile('build.gradle').write("""\
@@ -198,20 +199,20 @@ final class SimplePluginProject {
     return Files.createFile(file)
   }
 
-  String expectedReport = '''\
+  String expectedConsoleReport = '''\
     com.test.GreetingPlugin#apply(Ljava.lang.Object;)V ->
       com.test.GreetingPlugin#apply(Lorg.gradle.api.Project;)V ->
       org.gradle.api.Project#allprojects(Lorg.gradle.api.Action;)V
     
     com.test.GreetingPlugin#baz()V ->
       org.gradle.api.tasks.TaskContainer#getByName(Ljava.lang.String;)Lorg.gradle.api.Task;
- 
+    
     com.test.GreetingPlugin#baz()V ->
       org.gradle.api.tasks.TaskContainer#create(Ljava.lang.String;)Lorg.gradle.api.Task;
- 
+    
     com.test.GreetingPlugin#baz()V ->
       org.gradle.api.tasks.TaskContainer#all(Lorg.gradle.api.Action;)V
- 
+    
     com.test.GreetingPlugin#apply(Ljava.lang.Object;)V ->
       com.test.GreetingPlugin#apply(Lorg.gradle.api.Project;)V ->
       org.gradle.api.Project#getAllprojects()Ljava.util.Set;
@@ -244,7 +245,60 @@ final class SimplePluginProject {
       org.gradle.api.Project#subprojects(Lorg.gradle.api.Action;)V
   '''.stripIndent()
 
+  String expectedConsoleOutput = '''\
+      > Task :checkBestPractices FAILED
+      com.test.GreetingPlugin#apply(Ljava.lang.Object;)V ->
+        com.test.GreetingPlugin#apply(Lorg.gradle.api.Project;)V ->
+        org.gradle.api.Project#allprojects(Lorg.gradle.api.Action;)V
+
+      com.test.GreetingPlugin#baz()V ->
+        org.gradle.api.tasks.TaskContainer#getByName(Ljava.lang.String;)Lorg.gradle.api.Task;
+
+      com.test.GreetingPlugin#baz()V ->
+        org.gradle.api.tasks.TaskContainer#create(Ljava.lang.String;)Lorg.gradle.api.Task;
+
+      com.test.GreetingPlugin#baz()V ->
+        org.gradle.api.tasks.TaskContainer#all(Lorg.gradle.api.Action;)V
+
+      com.test.GreetingPlugin#apply(Ljava.lang.Object;)V ->
+        com.test.GreetingPlugin#apply(Lorg.gradle.api.Project;)V ->
+        org.gradle.api.Project#getAllprojects()Ljava.util.Set;
+
+      com.test.FancyTask#action()V ->
+        com.test.FancyTask#doAction()V ->
+        com.test.FancyTask$ReallyFancyTask#doAction()V ->
+        com.test.FancyTask$ReallyFancyTask#getProject()Lorg.gradle.api.Project;
+
+      com.test.ParentTask#action()V ->
+        com.test.ParentTask#foo()V ->
+        com.test.ParentTask#bar()V ->
+        com.test.ParentTask#doAction()V ->
+        com.test.ParentTask$ChildTask#doAction()V ->
+        com.test.ParentTask$ChildTask#getProject()Lorg.gradle.api.Project;
+
+      com.test.ParentTask2#action()V ->
+        com.test.ParentTask2#doAction()V ->
+        com.test.ParentTask2$ChildTask2#doAction()V ->
+        com.test.ParentTask2$ChildTask2#foo()V ->
+        com.test.ParentTask2$ChildTask2#bar([ILjava.lang.String;)V ->
+        com.test.ParentTask2$ChildTask2#getProject()Lorg.gradle.api.Project;
+
+      com.test.GreetingPlugin#apply(Ljava.lang.Object;)V ->
+        com.test.GreetingPlugin#apply(Lorg.gradle.api.Project;)V ->
+        org.gradle.api.Project#getSubprojects()Ljava.util.Set;
+
+      com.test.GreetingPlugin#apply(Ljava.lang.Object;)V ->
+        com.test.GreetingPlugin#apply(Lorg.gradle.api.Project;)V ->
+        org.gradle.api.Project#subprojects(Lorg.gradle.api.Action;)V
+
+      FAILURE: Build failed with an exception.
+    '''.stripIndent()
+
+  String expectedJsonReport = '''\
+    {"issues":[{"type":"allprojects","name":"allprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"allprojects","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"eager_api","name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"getByName","descriptor":"(Ljava/lang/String;)Lorg/gradle/api/Task;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"eager_api","name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"create","descriptor":"(Ljava/lang/String;)Lorg/gradle/api/Task;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"eager_api","name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"all","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_allprojects","name":"getAllprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"getAllprojects","descriptor":"()Ljava/util/Set;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_project","name":"getProject","trace":{"trace":[{"owner":"com/test/FancyTask","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/FancyTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/FancyTask$ReallyFancyTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/FancyTask$ReallyFancyTask","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_project","name":"getProject","trace":{"trace":[{"owner":"com/test/ParentTask","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"foo","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"bar","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask$ChildTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask$ChildTask","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_project","name":"getProject","trace":{"trace":[{"owner":"com/test/ParentTask2","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/ParentTask2","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"foo","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"bar","descriptor":"([ILjava/lang/String;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_subprojects","name":"getSubprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"getSubprojects","descriptor":"()Ljava/util/Set;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"subprojects","name":"subprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"subprojects","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}}]}
+  '''.stripIndent()
+
   String expectedBaseline = '''\
-    [{"name":"allprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"allprojects","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"getByName","descriptor":"(Ljava/lang/String;)Lorg/gradle/api/Task;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"create","descriptor":"(Ljava/lang/String;)Lorg/gradle/api/Task;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"all","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"getAllprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"getAllprojects","descriptor":"()Ljava/util/Set;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"getProject","trace":{"trace":[{"owner":"com/test/FancyTask","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/FancyTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/FancyTask$ReallyFancyTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/FancyTask$ReallyFancyTask","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"getProject","trace":{"trace":[{"owner":"com/test/ParentTask","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"foo","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"bar","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask$ChildTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask$ChildTask","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"getProject","trace":{"trace":[{"owner":"com/test/ParentTask2","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/ParentTask2","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"foo","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"bar","descriptor":"([ILjava/lang/String;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"getSubprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"getSubprojects","descriptor":"()Ljava/util/Set;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"name":"subprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"subprojects","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}}]
+    {"issues":[{"type":"allprojects","name":"allprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"allprojects","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"eager_api","name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"getByName","descriptor":"(Ljava/lang/String;)Lorg/gradle/api/Task;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"eager_api","name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"create","descriptor":"(Ljava/lang/String;)Lorg/gradle/api/Task;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"eager_api","name":"eagerApis","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"baz","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/tasks/TaskContainer","name":"all","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_allprojects","name":"getAllprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"getAllprojects","descriptor":"()Ljava/util/Set;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_project","name":"getProject","trace":{"trace":[{"owner":"com/test/FancyTask","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/FancyTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/FancyTask$ReallyFancyTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/FancyTask$ReallyFancyTask","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_project","name":"getProject","trace":{"trace":[{"owner":"com/test/ParentTask","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"foo","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"bar","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask$ChildTask","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask$ChildTask","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_project","name":"getProject","trace":{"trace":[{"owner":"com/test/ParentTask2","name":"action","descriptor":"()V","metadata":{"isTaskAction":true,"isVirtual":false}},{"owner":"com/test/ParentTask2","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"doAction","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"foo","descriptor":"()V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"bar","descriptor":"([ILjava/lang/String;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/ParentTask2$ChildTask2","name":"getProject","descriptor":"()Lorg/gradle/api/Project;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"get_subprojects","name":"getSubprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"getSubprojects","descriptor":"()Ljava/util/Set;","metadata":{"isTaskAction":false,"isVirtual":false}}]}},{"type":"subprojects","name":"subprojects","trace":{"trace":[{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Ljava/lang/Object;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"com/test/GreetingPlugin","name":"apply","descriptor":"(Lorg/gradle/api/Project;)V","metadata":{"isTaskAction":false,"isVirtual":false}},{"owner":"org/gradle/api/Project","name":"subprojects","descriptor":"(Lorg/gradle/api/Action;)V","metadata":{"isTaskAction":false,"isVirtual":false}}]}}]}
   '''.stripIndent()
 }

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/GradleBestPracticesExtension.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/GradleBestPracticesExtension.kt
@@ -28,7 +28,7 @@ open class GradleBestPracticesExtension @Inject constructor(
   objects: ObjectFactory
 ) {
 
-  internal val baseline = objects.fileProperty()
+  internal val baseline = objects.fileProperty().convention(layout.projectDirectory.file("best-practices-baseline.json"))
   internal val level = objects.property(Level::class.java).convention(Level.default)
 
   fun baseline(baseline: Provider<out RegularFile>) {
@@ -56,8 +56,6 @@ open class GradleBestPracticesExtension @Inject constructor(
   }
 
   internal companion object {
-
-    internal const val BASELINE_DEFAULT = "best-practices-baseline.json"
 
     internal fun create(project: Project) = project.extensions.create(
       "gradleBestPractices",

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/GradleBestPracticesPlugin.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/GradleBestPracticesPlugin.kt
@@ -1,8 +1,8 @@
 package com.autonomousapps
 
-import com.autonomousapps.GradleBestPracticesExtension.Companion.BASELINE_DEFAULT
 import com.autonomousapps.task.CheckBestPracticesTask
 import com.autonomousapps.task.CreateBaselineTask
+import com.autonomousapps.task.DetectBestPracticesViolationsTask
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.tasks.SourceSet
@@ -25,29 +25,41 @@ class GradleBestPracticesPlugin : Plugin<Project> {
 
       // A RegularFileProperty is allowed to wrap a nullable RegularFile
       @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
-      val bestPractices = tasks.register("checkBestPractices", CheckBestPracticesTask::class.java) {
+      val detectViolationsTask = tasks.register("detectViolations", DetectBestPracticesViolationsTask::class.java) {
         with(it) {
           classesDirs.setFrom(mainOutput)
           baseline.set(extension.baseline.map { f ->
             if (f.asFile.exists()) f else null
           })
-          creatingBaseline.set(provider { gradle.taskGraph.hasTask(baselineTask.get()) })
-          projectPath.set(project.path)
           logLevel.set(extension.level)
           outputJson.set(layout.buildDirectory.file("reports/best-practices/report.json"))
           outputText.set(layout.buildDirectory.file("reports/best-practices/report.txt"))
         }
       }
 
+      // A RegularFileProperty is allowed to wrap a nullable RegularFile
+      @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+      val checkBestPracticesTask = tasks.register("checkBestPractices", CheckBestPracticesTask::class.java) { t ->
+        with(t) {
+          reportJson.set(detectViolationsTask.flatMap { it.outputJson })
+          reportText.set(detectViolationsTask.flatMap { it.outputText })
+          baseline.set(extension.baseline.map { f ->
+            if (f.asFile.exists()) f else null
+          })
+          logLevel.set(extension.level)
+          projectPath.set(project.path)
+        }
+      }
+
       baselineTask.configure { t ->
         with(t) {
-          baseline.set(extension.baseline.orElse(layout.projectDirectory.file(BASELINE_DEFAULT)))
-          report.set(bestPractices.flatMap { it.outputJson })
+          baseline.set(extension.baseline)
+          report.set(detectViolationsTask.flatMap { it.outputJson })
         }
       }
 
       tasks.named("check").configure {
-        it.dependsOn(bestPractices)
+        it.dependsOn(checkBestPracticesTask)
       }
     }
   }

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/internal/utils/Json.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/internal/utils/Json.kt
@@ -34,8 +34,4 @@ internal object Json {
   inline fun <reified T> String.fromJson(): T {
     return getJsonAdapter<T>().fromJson(this)!!
   }
-
-  inline fun <reified T> String.fromJsonList(): List<T> {
-    return getJsonListAdapter<T>().fromJson(this)!!
-  }
 }

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/issue/Issue.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/issue/Issue.kt
@@ -3,6 +3,8 @@ package com.autonomousapps.issue
 import com.squareup.moshi.JsonClass
 import dev.zacsweers.moshix.sealed.annotations.TypeLabel
 
+class IssuesReport(val issues: List<Issue>)
+
 @JsonClass(generateAdapter = false, generator = "sealed:type")
 sealed class Issue {
   abstract val name: String

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/task/CheckBestPracticesTask.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/task/CheckBestPracticesTask.kt
@@ -1,35 +1,17 @@
 package com.autonomousapps.task
 
-import com.autonomousapps.internal.analysis.AllProjectsListener
-import com.autonomousapps.internal.analysis.ClassAnalyzer
-import com.autonomousapps.internal.analysis.CompositeIssueListener
-import com.autonomousapps.internal.analysis.EagerApisListener
-import com.autonomousapps.internal.analysis.GetAllprojectsListener
-import com.autonomousapps.internal.analysis.GetProjectListener
-import com.autonomousapps.internal.analysis.GetSubprojectsListener
-import com.autonomousapps.internal.analysis.IssueListener
-import com.autonomousapps.internal.analysis.SubprojectsListener
-import com.autonomousapps.internal.asm.ClassReader
 import com.autonomousapps.internal.logging.ConfigurableLogger
-import com.autonomousapps.internal.utils.Json.fromJsonList
-import com.autonomousapps.internal.utils.Json.toJson
-import com.autonomousapps.internal.utils.filterToClassFiles
-import com.autonomousapps.internal.utils.getAndDelete
-import com.autonomousapps.issue.Issue
-import com.autonomousapps.issue.IssueRenderer
+import com.autonomousapps.internal.utils.Json.fromJson
+import com.autonomousapps.issue.IssuesReport
 import org.gradle.api.DefaultTask
 import org.gradle.api.GradleException
-import org.gradle.api.file.ConfigurableFileCollection
 import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.logging.Logging
 import org.gradle.api.plugins.JavaBasePlugin
 import org.gradle.api.provider.Property
-import org.gradle.api.tasks.CacheableTask
 import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.InputFile
-import org.gradle.api.tasks.InputFiles
 import org.gradle.api.tasks.Optional
-import org.gradle.api.tasks.OutputFile
 import org.gradle.api.tasks.PathSensitive
 import org.gradle.api.tasks.PathSensitivity
 import org.gradle.api.tasks.TaskAction
@@ -38,7 +20,6 @@ import org.gradle.workers.WorkParameters
 import org.gradle.workers.WorkerExecutor
 import javax.inject.Inject
 
-@CacheableTask
 abstract class CheckBestPracticesTask @Inject constructor(
   private val workerExecutor: WorkerExecutor
 ) : DefaultTask() {
@@ -49,16 +30,17 @@ abstract class CheckBestPracticesTask @Inject constructor(
   }
 
   @get:PathSensitive(PathSensitivity.NONE)
-  @get:InputFiles
-  abstract val classesDirs: ConfigurableFileCollection
+  @get:InputFile
+  abstract val reportJson: RegularFileProperty
+
+  @get:PathSensitive(PathSensitivity.NONE)
+  @get:InputFile
+  abstract val reportText: RegularFileProperty
 
   @get:Optional
-  @get:PathSensitive(PathSensitivity.RELATIVE)
+  @get:PathSensitive(PathSensitivity.NONE)
   @get:InputFile
   abstract val baseline: RegularFileProperty
-
-  @get:Input
-  abstract val creatingBaseline: Property<Boolean>
 
   @get:Input
   abstract val projectPath: Property<String>
@@ -66,33 +48,23 @@ abstract class CheckBestPracticesTask @Inject constructor(
   @get:Input
   abstract val logLevel: Property<ConfigurableLogger.Level>
 
-  @get:OutputFile
-  abstract val outputJson: RegularFileProperty
-
-  @get:OutputFile
-  abstract val outputText: RegularFileProperty
-
   @TaskAction
   fun action() {
     workerExecutor.noIsolation().submit(Action::class.java) {
-      it.classesDirs.setFrom(classesDirs)
-      it.baseline.set(baseline)
-      it.creatingBaseline.set(creatingBaseline)
-      it.projectPath.set(projectPath)
-      it.logLevel.set(logLevel)
-      it.outputJson.set(outputJson)
-      it.outputText.set(outputText)
+      it.reportJson.set(this@CheckBestPracticesTask.reportJson)
+      it.reportText.set(this@CheckBestPracticesTask.reportText)
+      it.baseline.set(this@CheckBestPracticesTask.baseline)
+      it.projectPath.set(this@CheckBestPracticesTask.projectPath)
+      it.logLevel.set(this@CheckBestPracticesTask.logLevel)
     }
   }
 
   interface Parameters : WorkParameters {
-    val classesDirs: ConfigurableFileCollection
+    val reportJson: RegularFileProperty
+    val reportText: RegularFileProperty
     val baseline: RegularFileProperty
-    val creatingBaseline: Property<Boolean>
     val projectPath: Property<String>
     val logLevel: Property<ConfigurableLogger.Level>
-    val outputJson: RegularFileProperty
-    val outputText: RegularFileProperty
   }
 
   abstract class Action : WorkAction<Parameters> {
@@ -101,98 +73,32 @@ abstract class CheckBestPracticesTask @Inject constructor(
       ConfigurableLogger(this, parameters.logLevel.get())
     }
 
-    private val projectPath by lazy {
-      val path = parameters.projectPath.get()
-      if (path == ":") "" else path
-    }
-
-    private val baselineFixText by lazy {
-      "`./gradlew $projectPath:bestPracticesBaseline`"
-    }
-
     override fun execute() {
-      val outputJson = parameters.outputJson.getAndDelete()
-      val outputText = parameters.outputText.getAndDelete()
+      // The project path, unless it's ":", in which case use an empty string
+      val projectPath = parameters.projectPath.get().takeUnless { it == ":" } ?: ""
+      val baselineFixText = "`./gradlew $projectPath:bestPracticesBaseline`"
 
-      val classFiles = parameters.classesDirs.asFileTree.filterToClassFiles().files
-      logger.debug("classFiles=${classFiles.joinToString(prefix = "[", postfix = "]")}")
-
-      val issueListener = compositeListener()
-
-      // Visit every class file. Extract information into `issueListener`.
-      classFiles.forEach { classFile ->
-        classFile.inputStream().use { fis ->
-          ClassReader(fis.readBytes()).let { classReader ->
-            ClassAnalyzer(issueListener, logger).apply {
-              classReader.accept(this, 0)
-            }
-          }
-        }
-      }
-
-      // This does a global analysis, so must come after the forEach.
-      val issues = issueListener.computeIssues().sortedBy {
-        it.javaClass.canonicalName ?: it.javaClass.simpleName
-      }
-
+      val report = parameters.reportJson.get().asFile
+      val issues = report.readText().fromJson<IssuesReport>().issues
+      val text = parameters.reportText.get().asFile.readText()
       // Get baseline, if it exists.
-      val baseline = parameters.baseline.orNull?.asFile?.readText()?.fromJsonList<Issue>().orEmpty()
-      var hasNewIssues = false
-      var hasFixedIssues = false
+      val baseline = parameters.baseline.orNull?.asFile?.readText()?.fromJson<IssuesReport>()?.issues.orEmpty()
 
-      // Build console text.
-      val text = if (baseline.isEmpty()) {
-        // no baseline
-        issues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) }
-      } else {
-        // If we have a baseline, the behavior changes
-        // If we find an issue that isn't in the baseline, it's a new issue.
-        val newIssues = issues.filter { it !in baseline }
-        // any issue in the baseline that ISN'T also in the list of current issues has been fixed.
-        val fixedIssues = baseline.filter { it !in issues }
-        // any issue in the baseline that IS in the list of current issues is unfixed.
-        val unfixedIssues = baseline.filter { it in issues }
-        hasNewIssues = newIssues.isNotEmpty()
-        hasFixedIssues = fixedIssues.isNotEmpty()
+      // If we have a baseline, the behavior changes.
+      // If we find an issue that isn't in the baseline, it's a new issue.
+      val newIssues = issues.filter { it !in baseline }
+      // any issue in the baseline that ISN'T also in the list of current issues has been fixed.
+      val fixedIssues = baseline.filter { it !in issues }
+      val hasNewIssues = newIssues.isNotEmpty()
+      val hasFixedIssues = fixedIssues.isNotEmpty()
 
-        buildString {
-          if (newIssues.isNotEmpty()) {
-            appendLine("There are new issues:")
-            appendLine(newIssues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) })
-            appendLine()
-          } else {
-            appendLine("No new issues.")
-            appendLine()
-          }
-
-          if (fixedIssues.isNotEmpty()) {
-            appendLine("These issues have been resolved and should be removed from your baseline:")
-            appendLine(fixedIssues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) })
-            appendLine()
-          }
-
-          if (unfixedIssues.isNotEmpty()) {
-            appendLine("These issues have been ignored as part of your baseline:")
-            appendLine(unfixedIssues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) })
-            appendLine()
-          }
-        }
-      }
-
-      // Write output to disk.
-      outputText.writeText(text)
-      outputJson.writeText(issues.toJson())
-
-      // TODO maybe should split this task into two, one that computes issues and one that emits/throws.
       // Optionally print to console and throw exception.
-      if (parameters.creatingBaseline.get()) return
-
       if (issues.isNotEmpty()) {
         logger.report(text)
 
         if (baseline.isEmpty() || hasNewIssues) {
           val errorText = buildString {
-            appendLine("Violations of best practices detected. See the report at ${outputText.absolutePath} ")
+            appendLine("Violations of best practices detected. See the report at ${report.absolutePath} ")
             appendLine()
             appendLine("To create or update the baseline, run $baselineFixText")
           }
@@ -204,18 +110,6 @@ abstract class CheckBestPracticesTask @Inject constructor(
       if (hasFixedIssues) {
         throw GradleException("Your baseline contains resolved issues. Update with $baselineFixText")
       }
-    }
-
-    private fun compositeListener(): IssueListener {
-      val listeners = listOf(
-        AllProjectsListener(),
-        GetAllprojectsListener(),
-        SubprojectsListener(),
-        GetSubprojectsListener(),
-        GetProjectListener(),
-        EagerApisListener(),
-      )
-      return CompositeIssueListener(listeners)
     }
   }
 }

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/task/CreateBaselineTask.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/task/CreateBaselineTask.kt
@@ -6,7 +6,6 @@ import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.plugins.JavaBasePlugin
 import org.gradle.api.tasks.CacheableTask
 import org.gradle.api.tasks.InputFile
-import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.OutputFile
 import org.gradle.api.tasks.PathSensitive
 import org.gradle.api.tasks.PathSensitivity
@@ -24,7 +23,6 @@ abstract class CreateBaselineTask : DefaultTask() {
   @get:InputFile
   abstract val report: RegularFileProperty
 
-  @get:Optional
   @get:OutputFile
   abstract val baseline: RegularFileProperty
 

--- a/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/task/DetectBestPracticesViolationsTask.kt
+++ b/plugin-best-practices-plugin/src/main/kotlin/com/autonomousapps/task/DetectBestPracticesViolationsTask.kt
@@ -1,0 +1,169 @@
+package com.autonomousapps.task
+
+import com.autonomousapps.internal.analysis.AllProjectsListener
+import com.autonomousapps.internal.analysis.ClassAnalyzer
+import com.autonomousapps.internal.analysis.CompositeIssueListener
+import com.autonomousapps.internal.analysis.EagerApisListener
+import com.autonomousapps.internal.analysis.GetAllprojectsListener
+import com.autonomousapps.internal.analysis.GetProjectListener
+import com.autonomousapps.internal.analysis.GetSubprojectsListener
+import com.autonomousapps.internal.analysis.IssueListener
+import com.autonomousapps.internal.analysis.SubprojectsListener
+import com.autonomousapps.internal.asm.ClassReader
+import com.autonomousapps.internal.logging.ConfigurableLogger
+import com.autonomousapps.internal.utils.Json.fromJson
+import com.autonomousapps.internal.utils.Json.toJson
+import com.autonomousapps.internal.utils.filterToClassFiles
+import com.autonomousapps.internal.utils.getAndDelete
+import com.autonomousapps.issue.IssueRenderer
+import com.autonomousapps.issue.IssuesReport
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.logging.Logging
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.TaskAction
+import org.gradle.workers.WorkAction
+import org.gradle.workers.WorkParameters
+import org.gradle.workers.WorkerExecutor
+import javax.inject.Inject
+
+@CacheableTask
+abstract class DetectBestPracticesViolationsTask @Inject constructor(
+  private val workerExecutor: WorkerExecutor
+) : DefaultTask() {
+
+  @get:PathSensitive(PathSensitivity.NONE)
+  @get:InputFiles
+  abstract val classesDirs: ConfigurableFileCollection
+
+  @get:Optional
+  @get:PathSensitive(PathSensitivity.RELATIVE)
+  @get:InputFile
+  abstract val baseline: RegularFileProperty
+
+  @get:Input
+  abstract val logLevel: Property<ConfigurableLogger.Level>
+
+  @get:OutputFile
+  abstract val outputJson: RegularFileProperty
+
+  @get:OutputFile
+  abstract val outputText: RegularFileProperty
+
+  @TaskAction
+  fun action() {
+    workerExecutor.noIsolation().submit(Action::class.java) {
+      it.classesDirs.setFrom(classesDirs)
+      it.baseline.set(baseline)
+      it.logLevel.set(logLevel)
+      it.outputJson.set(outputJson)
+      it.outputText.set(outputText)
+    }
+  }
+
+  interface Parameters : WorkParameters {
+    val classesDirs: ConfigurableFileCollection
+    val baseline: RegularFileProperty
+    val logLevel: Property<ConfigurableLogger.Level>
+    val outputJson: RegularFileProperty
+    val outputText: RegularFileProperty
+  }
+
+  abstract class Action : WorkAction<Parameters> {
+
+    private val logger = Logging.getLogger(DetectBestPracticesViolationsTask::class.java.simpleName).run {
+      ConfigurableLogger(this, parameters.logLevel.get())
+    }
+
+    override fun execute() {
+      val outputJson = parameters.outputJson.getAndDelete()
+      val outputText = parameters.outputText.getAndDelete()
+
+      val classFiles = parameters.classesDirs.asFileTree.filterToClassFiles().files
+      logger.debug("classFiles=${classFiles.joinToString(prefix = "[", postfix = "]")}")
+
+      val issueListener = compositeListener()
+
+      // Visit every class file. Extract information into `issueListener`.
+      classFiles.forEach { classFile ->
+        classFile.inputStream().use { fis ->
+          ClassReader(fis.readBytes()).let { classReader ->
+            ClassAnalyzer(issueListener, logger).apply {
+              classReader.accept(this, 0)
+            }
+          }
+        }
+      }
+
+      // This does a global analysis, so must come after the forEach.
+      val issues = issueListener.computeIssues().sortedBy {
+        it.javaClass.canonicalName ?: it.javaClass.simpleName
+      }
+
+      // Get baseline, if it exists.
+      val baseline = parameters.baseline.orNull?.asFile?.readText()?.fromJson<IssuesReport>()?.issues.orEmpty()
+
+      // Build console text.
+      val text = if (baseline.isEmpty()) {
+        // no baseline
+        issues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) }
+      } else {
+        // If we have a baseline, the behavior changes
+        // If we find an issue that isn't in the baseline, it's a new issue.
+        val newIssues = issues.filter { it !in baseline }
+        // any issue in the baseline that ISN'T also in the list of current issues has been fixed.
+        val fixedIssues = baseline.filter { it !in issues }
+        // any issue in the baseline that IS in the list of current issues is unfixed.
+        val unfixedIssues = baseline.filter { it in issues }
+
+        buildString {
+          if (newIssues.isNotEmpty()) {
+            appendLine("There are new issues:")
+            appendLine(newIssues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) })
+            appendLine()
+          } else {
+            appendLine("No new issues.")
+            appendLine()
+          }
+
+          if (fixedIssues.isNotEmpty()) {
+            appendLine("These issues have been resolved and should be removed from your baseline:")
+            appendLine(fixedIssues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) })
+            appendLine()
+          }
+
+          if (unfixedIssues.isNotEmpty()) {
+            appendLine("These issues have been ignored as part of your baseline:")
+            appendLine(unfixedIssues.joinToString(separator = "\n\n") { IssueRenderer.renderIssue(it, pretty = true) })
+            appendLine()
+          }
+        }
+      }
+
+      // Write output to disk.
+      outputText.writeText(text)
+      outputJson.writeText(IssuesReport(issues).toJson())
+    }
+
+    private fun compositeListener(): IssueListener {
+      val listeners = listOf(
+        AllProjectsListener(),
+        GetAllprojectsListener(),
+        SubprojectsListener(),
+        GetSubprojectsListener(),
+        GetProjectListener(),
+        EagerApisListener(),
+      )
+      return CompositeIssueListener(listeners)
+    }
+  }
+}

--- a/settings.gradle
+++ b/settings.gradle
@@ -5,22 +5,35 @@ pluginManagement {
     gradlePluginPortal()
     mavenCentral()
   }
-  
+
   plugins {
-    id 'org.jetbrains.kotlin.jvm' version '1.7.20'
+    id 'com.gradle.enterprise' version '3.10.2'
     id 'com.gradle.plugin-publish' version '1.0.0'
-    id 'com.autonomousapps.plugin-best-practices-plugin' version '0.3'
-    id 'com.autonomousapps.dependency-analysis' version '1.15.0'
+    id 'com.autonomousapps.dependency-analysis' version '1.20.0'
+    id 'com.autonomousapps.plugin-best-practices-plugin' version '0.9'
+    id 'org.jetbrains.kotlin.jvm' version '1.9.0'
   }
+}
+
+plugins {
+  id 'com.gradle.enterprise'
 }
 
 rootProject.name = 'gradle-best-practices'
 
-include('plugin-best-practices-plugin')
+include 'plugin-best-practices-plugin'
 
 dependencyResolutionManagement {
   repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
   repositories {
     mavenCentral()
+  }
+}
+
+gradleEnterprise {
+  buildScan {
+    publishAlways()
+    termsOfServiceUrl = 'https://gradle.com/terms-of-service'
+    termsOfServiceAgree = 'yes'
   }
 }

